os/drivers/usbhost/uvc/usbhost_uvc_ctrl.c

```diff
@@ -0,0 +1,1360 @@
+	}
+
+	val = uvc_get_value(mapping, UVC_GET_CUR, control->cur);
+
+	if (mapping->v4l2_type == V4L2_CTRL_TYPE_MENU) {
+		menu = mapping->menu_info;
+
+		for (i = 0; i < mapping->menu_count; ++i, ++menu) {
+			if (menu->value == val) {
+				val = i;
+				break;
+			}
+		}
+	}
+
+	xctrl->value = val;
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: uvc_ctrl_set
+ *
+ * Description:
+ *   This function is used to set the control value to the camera device.
+ *   Will verify the range to which the value can be set on device, also backup
+ *   the previous values to restore when required.
+ *
+ *   This function is called from the uvc_set_ctrlvalue method of V4L2.
+ *
+ * Input Parameters:
+ *   chain - reference to the video chain object.
+ *   ctrl_class - class the control command to be set.
+ *   xctrl - control information to be set on device.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_set(struct uvc_video_chain *chain, uint16_t ctrl_class, struct v4l2_ext_control *xctrl)
+{
+	struct uvc_control *ctrl;
+	struct uvc_ctrl_value val;
+	struct uvc_control_mapping *mapping;
+	int32_t value;
+	uint32_t step;
+	int32_t min;
+	int32_t max;
+	int ret;
+
+	ctrl = uvc_search_control(chain, xctrl->id, ctrl_class, &mapping);
+	if (ctrl == NULL) {
+		udbg("Failed to find a valid control!!\n");
+		return -EINVAL;
+	}
+	if (!(ctrl->info.flags & F_SET_CUR)) {
+		uvdbg("Invalid access to control, not supported!!\n");
+		return -EACCES;
+	}
+
+	/* range check, value should be in the range of min and max resoluton of steps */
+	switch (mapping->v4l2_type) {
+	case V4L2_CTRL_TYPE_INTEGER:
+		ret = uvc_ctrl_populate_values(chain, ctrl);
+		if (ret < 0) {
+			return ret;
+		}
+
+		uvc_read_control(ctrl, mapping, &val);
+		min = val.min;
+		max = val.max;
+		step = val.step;
+		if (step == 0) {
+			step = 1;
+		}
+
+		CHECK_RANGE(xctrl->value, min, max, step);
+
+		xctrl->value = min + ((uint32_t)(xctrl->value - min) + step / 2)
+					   / step * step;
+		value = xctrl->value;
+		break;
+
+	case V4L2_CTRL_TYPE_BOOLEAN:
+		CHECK_RANGE(xctrl->value, 0, 1, 1);
+		value = xctrl->value;
+		break;
+
+	case V4L2_CTRL_TYPE_MENU:
+		if (xctrl->value < 0 || xctrl->value >= mapping->menu_count) {
+			return -ERANGE;
+		}
+		value = mapping->menu_info[xctrl->value].value;
+
+		/* Valid menu indices are reported by the GET_RES request for
+		 * UVC controls that support it.
+		 */
+		if (mapping->data_type == UVC_CTRL_DATA_TYPE_BITMASK && (ctrl->info.flags & F_GET_RES)) {
+			ret = uvc_ctrl_populate_values(chain, ctrl);
+			if (ret < 0) {
+				udbg("Failed to populate control values!!\n");
+				return ret;
+			}
+
+			step = uvc_get_value(mapping, UVC_GET_RES, ctrl->step);
+			if (!(step & value)) {
+				uvdbg("Failed to get setps from control\n");
+				return -EINVAL;
```

Comment:
EINVAL is not correct, In my opinion. Please check this once again.

Reply:
Other option which I can think of is EPERM, "operation for permitted". Corrected to EPERM currently.